File: frontend/src/types/market.ts
/**
 * 市场数据类型定义 (Market Data Types)
 * 
 * 文件作用：
 * - 定义所有市场数据相关的 TypeScript 类型
 * - 确保前后端数据结构一致性
 * - 提供完整的类型安全和智能提示
 * 
 * 类型分类：
 * 
 * 1. K线数据 (Candlestick Data)
 *    - KlineData: 单根K线的OHLCV数据
 *    - KlineResponse: API返回的K线数据响应
 * 
 * 2. 实时行情 (Ticker Data)
 *    - TickerData: 实时价格、涨跌幅、成交量等
 * 
 * 3. 交易对信息 (Symbol Info)
 *    - SymbolInfo: 交易对的基本信息
 *    - SymbolListResponse: 交易对列表响应
 * 
 * 4. 合约数据 (Contract Data)
 *    - FundingRateData: 资金费率
 *    - OpenInterestData: 持仓量
 * 
 * 5. 技术指标 (Technical Indicators)
 *    - IndicatorLatestValues: 最新指标值（EMA、MACD、RSI、ATR）
 *    - IndicatorSeriesData: 指标时间序列数据
 *    - IndicatorsResponse: API返回的指标响应
 * 
 * 6. 时间周期 (Time Interval)
 *    - TimeInterval: 支持的K线时间周期类型
 * 
 * 数据字段说明：
 * - timestamp: Unix时间戳（毫秒）
 * - open/high/low/close: OHLC价格
 * - volume: 成交量
 * - bid/ask: 买一价/卖一价
 * - change: 价格变化量
 * - percentage: 涨跌幅百分比
 * - funding_rate: 资金费率（正数=多头付费，负数=空头付费）
 * - open_interest: 未平仓合约数量
 * 
 * null 值说明：
 * - 某些字段可能为 null（如现货市场没有资金费率）
 * - 使用时需要做 null 检查
 * 
 * 使用示例：
 * const kline: KlineData = {
 *   timestamp: 1698765432000,
 *   open: 43250.5,
 *   high: 43500.0,
 *   low: 43200.0,
 *   close: 43450.0,
 *   volume: 123.45
 * };
 */

export interface KlineData {
  timestamp: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface KlineResponse {
  symbol: string;
  interval: string;
  data: KlineData[];
  count: number;
}

export interface TickerData {
  symbol: string;
  last: number;
  bid: number | null;
  ask: number | null;
  high: number | null;
  low: number | null;
  volume: number | null;
  change: number | null;
  percentage: number | null;
  timestamp: number;
}

export interface SymbolInfo {
  symbol: string;
  base: string;
  quote: string;
  active: boolean;
}

export interface SymbolListResponse {
  symbols: SymbolInfo[];
  count: number;
}

export interface FundingRateData {
  symbol: string;
  funding_rate: number | null;
  next_funding_time: number | null;
  timestamp: number;
}

export interface OpenInterestData {
  symbol: string;
  open_interest: number | null;
  timestamp: number;
}

export interface IndicatorLatestValues {
  ema20: number;
  ema50: number;
  macd: number;
  signal: number;
  histogram: number;
  rsi7: number;
  rsi14: number;
  atr3: number;
  atr14: number;
}

export interface IndicatorSeriesData {
  timestamps: number[];
  ema20: number[];
  ema50: number[];
  macd: number[];
  signal: number[];
  histogram: number[];
  rsi7: number[];
  rsi14: number[];
  atr3: number[];
  atr14: number[];
}

export interface IndicatorsResponse {
  symbol: string;
  interval: string;
  latest_values: IndicatorLatestValues;
  series_data?: IndicatorSeriesData;
}

export type TimeInterval = '1m' | '5m' | '15m' | '1h' | '4h' | '1d';

File: backend/app/schemas/market.py
"""
市场数据相关的 Pydantic schemas
"""
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime


class KlineData(BaseModel):
    """单根K线数据"""
    timestamp: int = Field(..., description="时间戳（毫秒）")
    open: float = Field(..., description="开盘价")
    high: float = Field(..., description="最高价")
    low: float = Field(..., description="最低价")
    close: float = Field(..., description="收盘价")
    volume: float = Field(..., description="成交量")
    
    class Config:
        json_schema_extra = {
            "example": {
                "timestamp": 1698768000000,
                "open": 34500.0,
                "high": 34800.0,
                "low": 34400.0,
                "close": 34650.0,
                "volume": 123.45
            }
        }


class KlineResponse(BaseModel):
    """K线数据响应"""
    symbol: str = Field(..., description="交易对")
    interval: str = Field(..., description="时间周期")
    data: List[KlineData] = Field(..., description="K线数据列表")
    count: int = Field(..., description="数据条数")
    
    class Config:
        json_schema_extra = {
            "example": {
                "symbol": "BTC/USDT",
                "interval": "1h",
                "data": [],
                "count": 100
            }
        }


class TickerData(BaseModel):
    """实时行情数据"""
    symbol: str = Field(..., description="交易对")
    last: float = Field(..., description="最新成交价")
    bid: Optional[float] = Field(None, description="买一价")
    ask: Optional[float] = Field(None, description="卖一价")
    high: Optional[float] = Field(None, description="24h最高价")
    low: Optional[float] = Field(None, description="24h最低价")
    volume: Optional[float] = Field(None, description="24h成交量")
    change: Optional[float] = Field(None, description="24h涨跌额")
    percentage: Optional[float] = Field(None, description="24h涨跌幅（%）")
    timestamp: int = Field(..., description="时间戳（毫秒）")
    
    class Config:
        json_schema_extra = {
            "example": {
                "symbol": "BTC/USDT",
                "last": 34650.0,
                "bid": 34649.5,
                "ask": 34650.5,
                "high": 35000.0,
                "low": 34000.0,
                "volume": 12345.67,
                "change": 650.0,
                "percentage": 1.91,
                "timestamp": 1698768000000
            }
        }


class SymbolInfo(BaseModel):
    """交易对信息"""
    symbol: str = Field(..., description="交易对")
    base: str = Field(..., description="基础货币")
    quote: str = Field(..., description="计价货币")
    active: bool = Field(..., description="是否活跃")
    
    class Config:
        json_schema_extra = {
            "example": {
                "symbol": "BTC/USDT",
                "base": "BTC",
                "quote": "USDT",
                "active": True
            }
        }


class SymbolListResponse(BaseModel):
    """交易对列表响应"""
    symbols: List[SymbolInfo] = Field(..., description="交易对列表")
    count: int = Field(..., description="交易对数量")


class MarketStats(BaseModel):
    """市场统计数据"""
    symbol: str = Field(..., description="交易对")
    price: float = Field(..., description="当前价格")
    high_24h: float = Field(..., description="24h最高价")
    low_24h: float = Field(..., description="24h最低价")
    volume_24h: float = Field(..., description="24h成交量")
    change_24h: float = Field(..., description="24h涨跌额")
    change_percentage_24h: float = Field(..., description="24h涨跌幅（%）")
    timestamp: int = Field(..., description="时间戳")

